exonum/src/runtime/configuration_new/mod.rs

```diff
@@ -124,6 +156,104 @@ impl ConfigurationService for ConfigurationServiceImpl {
+
+        let dispatcher = self.dispatcher.borrow();
+        dispatcher.start_deploy(artifact_spec).map_err(|err| {
+            error!("Service instance deploy failed: {:?}", err);
+            ServiceError::DeployError(err)
+        })?;
+
+        // TODO add result into deployable (to check deploy status in before_commit).
+
+        Ok(())
+    }
+
+    fn init(
+        &self,
+        mut ctx: TransactionContext,
+        arg: transactions::Init,
+    ) -> Result<(), ExecutionError> {
+        let artifact_spec = arg.get_artifact_spec();
+
+        let dispatcher = self.dispatcher.borrow();
+
+        let instance_id = self
+            .assign_service_id(ctx.fork(), &arg.instance_name)
+            .ok_or(ServiceError::ServiceInstanceNameInUse)?;
+
+        let init_data = InstanceInitData {
+            instance_id,
+            constructor_data: arg.constructor_data,
+        };
+
+        dispatcher
+            .init_service(ctx.env_context(), artifact_spec, &init_data)
+            .map_err(|err| {
+                error!("Service instance initialization failed: {:?}", err);
+                ServiceError::InitError(err)
+            })?;
+
+        Ok(())
+    }
+
+    fn deploy_and_init(
+        &self,
+        mut ctx: TransactionContext,
+        arg: transactions::DeployInit,
+    ) -> Result<(), ExecutionError> {
+        // TODO reduce copy-paste.
+
+        // Deploy
+        {
+            let artifact_spec = arg.deploy_tx.get_artifact_spec();
+
+            let dispatcher = self.dispatcher.borrow();
+            dispatcher.start_deploy(artifact_spec.clone()).map_err(|err| {
+                error!("Service instance deploy failed: {:?}", err);
+                ServiceError::DeployError(err)
+            })?;
+
+            // Check if service is deployed.
+            let cancel_if_incomplete = true;
+            if dispatcher.check_deploy_status(artifact_spec, cancel_if_incomplete).map_err(|err| ServiceError::DeployError(err))?
+                != DeployStatus::Deployed
+            {
+                return Err(ServiceError::InitError(InitError::NotDeployed))?;
+            }
+        }
+
+        // Init
+        {
+            let artifact_spec = arg.init_tx.get_artifact_spec();
+
+            let dispatcher = self.dispatcher.borrow();
+
+            let instance_id = self
+                .assign_service_id(ctx.fork(), &arg.init_tx.instance_name)
+                .ok_or(ServiceError::ServiceInstanceNameInUse)?;
+
+            let init_data = InstanceInitData {
+                instance_id,
+                constructor_data: arg.init_tx.constructor_data,
+            };
+
+            dispatcher
+                .init_service(ctx.env_context(), artifact_spec, &init_data)
+                .map_err(|err| {
+                    error!("Service instance initialization failed: {:?}", err);
+                    ServiceError::InitError(err)
+                })?;
+        }
+
+        Ok(())
+    }
 }
```

Comment:
It seems to me that a code duplication here, doesn't it ? 🙃

Reply:
Yep, it is. And there is a comment for that slightly above: `// TODO reduce copy-paste.`. However, it can't be fixed easily, since `TransactionContext` is consumed in tx handling and doesn't implement `Clone`. 

Since it's a draft I decided to left this thing as-is for now and move to the more important parts of this task.